### PR TITLE
Add XPath function checksum

### PR DESCRIPTION
**Pull request by shubham1g5**
Case url - https://manage.dimagi.com/default.asp?255290#1352684

Implements checksum by Verhoeff Algorithm. Code based on implementation here - https://en.wikibooks.org/wiki/Algorithm_Implementation/Checksums/Verhoeff_Algorithm#Java

### CODE REVIEW

File: src/main/java/org/commcare/cases/util/StringUtils.java [hunk 1]
```diff
@@ -134,4 +135,18 @@ private static int LevenshteinDistance(String s0, String s1) {
...
+
+    /**
+     * Converts a string to a string array.
+     */
+    public static String[] toArray(String str) {
+
+        String[] myArray = new String[str.length()];
+
+        for (int i = 0; i < str.length(); i++) {
+            myArray[i] = str.substring(i, i + 1);
```
Comment (wpride) [inline src/main/java/org/commcare/cases/util/StringUtils.java]: This might be more succinct using the String `charAt()` function (and possibly this whole function should return a `char[]` rather than a `String[]`)

File: src/main/java/org/javarosa/core/util/ArrayUtilities.java [hunk 2]
```diff
@@ -33,6 +35,20 @@ public static boolean arraysEqual(byte[] array1, byte[] array2) {
     }
 
     /**
+     * Reverses an Array
+     */
+    public static <T> T[] reverse(Object[] myArray, Class<? extends T[]> tClass) {
```
Comment (wpride) [inline src/main/java/org/javarosa/core/util/ArrayUtilities.java]: If you switch to using a `char[]` you could use Apache Commons `ArrayUtils.reverse` https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/ArrayUtils.html#reverse-char:A-

File: src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        name = NAME;
+        expectedArgCount = EXPECTED_ARG_COUNT;
+    }
+
+    public XPathChecksumFunc(XPathExpression[] args) throws XPathSyntaxException {
+        super(NAME, args, EXPECTED_ARG_COUNT, true);
+    }
+
+    @Override
+    public Object evalBody(DataInstance model, EvaluationContext evalContext, Object[] evaluatedArgs) {
```
Comment (wpride) [inline src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java]: can be static

File: src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        return checksum(evaluatedArgs[0], evaluatedArgs[1]);
+    }
+
+    /**
+     * @param o1 algorithm type used to calculate checksum. We only support 'verhoeff' for now.
+     * @param o2 input we are calculating checksum for
+     * @return checksum of {@code o2} calculated using  {@code o1} type algorithm
+     */
+    private String checksum(Object o1, Object o2) {
```
Comment (wpride) [inline src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java]: can be static

File: src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * Calculates Verhoeff checksum for the given {@code input} string <p>
+     *
+     * @param input input string to calculate verhoeff checksum for
+     * @return Verhoeff checksum value for {@code input}
+     * @see <a href="https://en.wikibooks.org/wiki/Algorithm_Implementation/Checksums/Verhoeff_Algorithm#Java">Based on Verhoeff checksum implementation here</a>
+     */
+    private String verhoeffChecksum(String input) {
```
Comment (wpride) [inline src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java]: can be static

File: src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
+package org.javarosa.xpath.expr;
+
+import org.commcare.cases.util.StringUtils;
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.instance.DataInstance;
+import org.javarosa.core.util.ArrayUtilities;
+import org.javarosa.xpath.XPathUnsupportedException;
+import org.javarosa.xpath.parser.XPathSyntaxException;
+
+public class XPathChecksumFunc extends XPathFuncExpr {
```
Comment (wpride) [inline src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java]: Good looking class!

File: src/main/java/org/javarosa/core/util/ArrayUtilities.java (hunk 2 above)
Comment (shubham1g5) [inline src/main/java/org/javarosa/core/util/ArrayUtilities.java]: @wpride It seems like currently we are not using 'commons-lang3' library in the project. Should I go ahead and add it.

File: src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java (hunk 5 above)
Comment (shubham1g5) [inline src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java]: 👍

File: src/main/java/org/commcare/cases/util/StringUtils.java (hunk 1 above)
Comment (shubham1g5) [inline src/main/java/org/commcare/cases/util/StringUtils.java]: 👍

File: src/main/java/org/javarosa/core/util/ArrayUtilities.java (hunk 2 above)
Comment (wpride) [inline src/main/java/org/javarosa/core/util/ArrayUtilities.java]: Usually I prefer adding a dependency to re-implementing functionality but definitely your call

Comment (amstone326) [inline src/main/java/org/javarosa/core/util/ArrayUtilities.java]: Another option is if you start with splitting your `input` String to the `verhoeffChecksum` function into an `List` instead of an array, then you can use `Collections.reverse()`

File: src/test/java/org/javarosa/xpath/test/XPathEvalTest.java [hunk 7]
```diff
@@ -58,8 +58,8 @@ private void testEval(String expr, FormInstance model, EvaluationContext ec, Obj
             if (exceptionExpected) {
                 fail("Expected exception, expression : " + expr);
             } else if ((result instanceof Double && expected instanceof Double)) {
-                Double o = (Double)result;
-                Double t = (Double)expected;
+                Double o = (Double) result;
```
Comment (amstone326) [inline src/test/java/org/javarosa/xpath/test/XPathEvalTest.java]: So I cannot currently find the style guide of ours that I am pretty sure exists, but our practice with casting is actually to leave no space between the closing paren and the target of the cast. Sometimes Android Studio tries to change this for you automatically which is incredibly annoying, so I'm not even sure if you changed these explicitly, but would you mind switching back the lines in this file that were changed? (Also note that I am a big fan of all other linting changes you made in this PR 😄)

File: src/test/java/org/javarosa/xpath/test/XPathEvalTest.java [hunk 8]
```diff
@@ -523,6 +523,10 @@ public void doTests() {
         testEval("id-compress(0, 'CD','','ABCDE',1)", null, ec, new XPathException());
         testEval("id-compress(0, 'CD','CD','ABCDE',1)", null, ec, new XPathException());
 
+        testEval("checksum('verhoeff','41310785898')", null, null, "4");
```
Comment (amstone326) [inline src/test/java/org/javarosa/xpath/test/XPathEvalTest.java]: Can you remind me how you determined the correct values of these outputs prior to running them in the code here?

File: src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    private String checksum(Object o1, Object o2) {
+        String algorithmKey = FunctionUtils.toString(o1);
+        String input = FunctionUtils.toString(o2);
+
+        switch (algorithmKey) {
+            case ALGORITHM_KEY_VERHOEFF:
+                return verhoeffChecksum(input);
+            default:
+                throw new XPathUnsupportedException("Algorithm key " + algorithmKey + " is not supported for checksum");
```
Comment (shubham1g5) [inline src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java]: 👍

File: src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                {9, 4, 5, 3, 1, 2, 6, 8, 7, 0},
+                {4, 2, 8, 6, 5, 7, 3, 9, 0, 1},
+                {2, 7, 9, 3, 8, 0, 6, 4, 1, 5},
+                {7, 0, 4, 6, 9, 1, 3, 2, 5, 8}
+        };
+
+        // The inverse table
+        int[] inv = {0, 4, 3, 2, 1, 5, 6, 7, 8, 9};
+
+        int check = 0;
```
Comment (shubham1g5) [inline src/main/java/org/javarosa/xpath/expr/XPathChecksumFunc.java]: 👍

File: src/test/java/org/javarosa/xpath/test/XPathEvalTest.java (hunk 7 above)
Comment (shubham1g5) [inline src/test/java/org/javarosa/xpath/test/XPathEvalTest.java]: Yeah. All the formatting changes are due to Android Studio auto format. I think we really need to have a standard style.xml that we can import in Android Studio and which Android Studio can then utilise to format/style the code to avoid these conflicts. Though for now will revert these changes.

File: src/test/java/org/javarosa/xpath/test/XPathEvalTest.java (hunk 8 above)
Comment (shubham1g5) [inline src/test/java/org/javarosa/xpath/test/XPathEvalTest.java]: I took some sample Aadhar ids and passed the first 11 digits to the checksum function while the last digit as the expected outcome. So for this case Aadhar Id I used would be  '413107858984".

File: src/test/java/org/javarosa/xpath/test/XPathEvalTest.java (hunk 7 above)
Comment (amstone326) [inline src/test/java/org/javarosa/xpath/test/XPathEvalTest.java]: Agreed. For now I found how to configure AS to stop adding a space after casts, so will show you that today

File: src/main/java/org/javarosa/core/util/ArrayUtilities.java (hunk 2 above)
Comment (shubham1g5) [inline src/main/java/org/javarosa/core/util/ArrayUtilities.java]: 👍

**Comment by shubham1g5** [comment]
Pushed all the suggested changes. @amstone326  Thanks for catching the error in FunctionUtils. Weird that it didn't fail on Jenkins build.

File: src/test/java/org/javarosa/xpath/test/XPathEvalTest.java (hunk 8 above)
Comment (amstone326) [inline src/test/java/org/javarosa/xpath/test/XPathEvalTest.java]: Ah perfect, I forgot that property of the Adhaar IDs is the whole point
